File: poses_viewer/src/drill2d/__tests__/repFilter.test.ts
import { describe, expect, it } from 'vitest'
import { filterReps } from '../repFilter'
import { Stroke2D } from '../types'

/** Mirrors RepFilterTest.kt. */
const stroke = (i: number, peakSpeed: number, durFrames: number): Stroke2D => ({
  strokeIndex: i,
  startFrame: i * 30,
  peakFrame: i * 30 + Math.floor(durFrames / 2),
  endFrame: i * 30 + durFrames,
  peakSpeed,
})

describe('filterReps (mirrors RepFilterTest)', () => {
  it('uniform strokes all kept', () => {
    const s = Array.from({ length: 6 }, (_, i) => stroke(i, 2.4, 6))
    expect(filterReps(s)).toEqual(s)
  })

  it('slow and overlong junk dropped', () => {
    const good = Array.from({ length: 6 }, (_, i) => stroke(i, 2.4, 6))
    const slow = stroke(6, 1.1, 6)   // ball pickup: above detector threshold, half the cluster speed
    const smear = stroke(7, 2.4, 20) // walking: long movement, plausible peak
    expect(filterReps([...good, slow, smear])).toEqual(good)
  })

  it('too few strokes are not filtered', () => {
    const s = [stroke(0, 2.4, 6), stroke(1, 0.5, 30), stroke(2, 5, 2)]
    expect(filterReps(s)).toEqual(s)
  })
})

File: shared/src/commonTest/kotlin/com/ttcoachai/shared/drill/RepFilterTest.kt
package com.ttcoachai.shared.drill

import com.ttcoachai.shared.models.Stroke2D
import kotlin.test.Test
import kotlin.test.assertEquals

class RepFilterTest {

    private fun stroke(i: Int, peakSpeed: Float, durFrames: Int) = Stroke2D(
        strokeIndex = i,
        startFrame = i * 30,
        peakFrame = i * 30 + durFrames / 2,
        endFrame = i * 30 + durFrames,
        peakSpeed = peakSpeed
    )

    @Test
    fun uniformStrokesAllKept() {
        val s = List(6) { stroke(it, 2.4f, 6) }
        assertEquals(s, RepFilter.filter(s))
    }

    @Test
    fun slowAndOverlongJunkDropped() {
        val good = List(6) { stroke(it, 2.4f, 6) }
        val slow = stroke(6, 1.1f, 6)    // ball pickup: above detector threshold, half the cluster speed
        val smear = stroke(7, 2.4f, 20)  // walking: long movement, plausible peak
        assertEquals(good, RepFilter.filter(good + slow + smear))
    }

    @Test
    fun tooFewStrokesAreNotFiltered() {
        val s = listOf(stroke(0, 2.4f, 6), stroke(1, 0.5f, 30), stroke(2, 5f, 2))
        assertEquals(s, RepFilter.filter(s), "below 4 strokes there is no cluster to trust")
    }
}
